packages/preview3-shim/lib/nodejs/filesystem/descriptor.js

```diff
@@ -0,0 +1,976 @@
+                throw new FsError(
+                    'invalid',
+                    `Unknown new-timestamp tag: ${newTimestamp.tag}`
+                );
+        }
+    }
+
+    async #computeTimestamps(atimeDesc, mtimeDesc, path = null) {
+        let stats;
+        if (atimeDesc.tag === 'no-change' || mtimeDesc.tag === 'no-change') {
+            const flags = { symlinkFollow: true };
+            const p = path === null ? this.stat() : this.statAt(flags, path);
+            stats = await p;
+        }
+
+        const atime = this.#getNewTimestamp(
+            atimeDesc,
+            stats?.dataAccessTimestamp
+        );
+        const mtime = this.#getNewTimestamp(
+            mtimeDesc,
+            stats?.dataModificationTimestamp
+        );
+        return { atime, mtime };
+    }
+
+    /**
+     * Resolve a relative subpath against this descriptor.
+     */
+    #getFullPath(subpath, _followSymlinks) {
+        subpath = subpath.replaceAll('\\', '/').replace(/\/\/+/g, '/');
+
+        if (subpath.startsWith('/')) {
+            throw new FsError('not-permitted');
+        }
+
+        const segments = [];
+        for (const seg of subpath.split('/')) {
+            if (seg === '' || seg === '.') {
+                continue;
+            }
+            if (seg === '..') {
+                if (segments.length === 0) {
+                    throw new FsError('not-permitted');
+                }
+                segments.pop();
+            } else {
+                segments.push(seg);
+            }
+        }
+
+        const base = this.#hostPreopen ?? this.#fullPath;
+        if (segments.length === 0) {
+            return base;
+        }
+
+        const baseNormalized = stripTrailingSlash(base);
+        return `${baseNormalized}/${segments.join('/')}`;
+    }
+
+    #ensureHandle() {
+        if (!this.#handle) {
+            throw new FsError('bad-descriptor');
+        }
+    }
+}
+
+function lookupType(obj) {
+    if (obj.isFile()) return 'regular-file';
+    else if (obj.isSocket()) return 'socket';
+    else if (obj.isSymbolicLink()) return 'symbolic-link';
+    else if (obj.isFIFO()) return 'fifo';
+    else if (obj.isDirectory()) return 'directory';
+    else if (obj.isCharacterDevice()) return 'character-device';
+    else if (obj.isBlockDevice()) return 'block-device';
+    return 'unknown';
+}
+
+const NS_PER_SEC = 1_000_000_000n;
+function nsToDateTime(ns) {
+    const seconds = ns / NS_PER_SEC;
+    const nanoseconds = Number(ns % NS_PER_SEC);
+    return { seconds, nanoseconds };
+}
+
+function timestampToSec({ seconds, nanoseconds }) {
+    return Number(seconds) + nanoseconds / 1e9;
+}
+
+function stripTrailingSlash(path) {
+    if (path === '/') return '/';
+    return path.replace(/\/+$/, '');
+}
+
+const descriptorCreatePreopen = Descriptor._createPreopen;
+delete Descriptor._createPreopen;
+const descriptorCreate = Descriptor._create;
+delete Descriptor._create;
+
+export const types = { Descriptor };
```

Comment:
```suggestion
export const types = { Descriptor };
```

Could this be removed? Naming exports `types` seems a bit weird, does something need access to the class directly except for via `descriptorCreate`/`descriptorCreatePreopen`?